pkg/sign/signer.go

```diff
@@ -15,29 +15,218 @@
+	}
+
+	jwtString, err := base64.RawURLEncoding.DecodeString(tokenParts[1])
+	if err != nil {
+		return nil, err
+	}
+
+	var jwt jsonWebToken
+	err = json.Unmarshal([]byte(jwtString), &jwt)
+	if err != nil {
+		return nil, err
+	}
+
+	// Generate ephemeral keypair
+	privateKey, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
+	if err != nil {
+		return nil, err
+	}
+
+	pubKeyBytes, err := x509.MarshalPKIXPublicKey(privateKey.Public())
+	if err != nil {
+		return nil, err
+	}
+
+	pemBlock := pem.Block{
+		Type:  "PUBLIC KEY",
+		Bytes: pubKeyBytes,
+	}
+
+	var pemBytes bytes.Buffer
+	if err = pem.Encode(&pemBytes, &pemBlock); err != nil {
+		return nil, err
+	}
+
+	// Sign JWT subject with ephemeral key for proof of possession
+	subjectDigest := sha256.Sum256([]byte(jwt.Sub))
+	subjectSignature, err := privateKey.Sign(rand.Reader, subjectDigest[:], nil)
+	if err != nil {
+		return nil, err
+	}
+
+	// Make Fulcio certificate request
+	certRequest := fulcioCertRequest{
+		Credentials: identityToken{
+			OIDCIdentityToken: f.options.IdentityToken,
+		},
+		PublicKeyRequest: publicKeyRequest{
+			PublicKey: publicKey{
+				Algorithm: "ECDSA",
+				Content:   pemBytes.String(),
+			},
+			ProofOfPossession: base64.StdEncoding.EncodeToString(subjectSignature),
+		},
+	}
 
-// func (f *Fulcio) Sign(data []byte) (*protobundle.Bundle, error) {}
+	requestJSON, err := json.Marshal(&certRequest)
+	if err != nil {
+		return nil, err
+	}
+	requestBytes := bytes.NewBuffer(requestJSON)
+
+	var client http.Client
+	if f.options.Timeout != 0 {
+		client.Timeout = f.options.Timeout
+	}
+
+	request, err := http.NewRequest("POST", f.options.BaseURL+"/api/v2/signingCert", requestBytes)
+	if err != nil {
+		return nil, err
+	}
+	request.Header.Add("Authorization", "Bearer "+f.options.IdentityToken)
+	request.Header.Add("Content-Type", "application/json")
+	request.Header.Add("User-Agent", constructUserAgent(f.options.LibraryVersion))
+
+	response, err := client.Do(request)
+	if err != nil {
+		return nil, err
+	}
+
+	body, err := io.ReadAll(response.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	if response.StatusCode != 200 {
+		return nil, fmt.Errorf("Fulcio returned %d: %s", response.StatusCode, string(body))
+	}
+
+	// Assemble bundle from Fulcio response
+	var fulcioResp fulcioResponse
+	err = json.Unmarshal(body, &fulcioResp)
+	if err != nil {
+		return nil, err
+	}
+
+	certs := fulcioResp.SCT.Chain.Certificates
+	if len(certs) == 0 {
+		return nil, errors.New("Fulcio returned no certificates")
+	}
+
+	certBlock, _ := pem.Decode([]byte(certs[0]))
+	if certBlock == nil {
+		return nil, errors.New("unable to parse Fulcio certificate")
+	}
+
+	data := content.PreAuthEncoding()
+	dataDigest := sha256.Sum256([]byte(data))
+	signature, err := ecdsa.SignASN1(rand.Reader, privateKey, dataDigest[:])
+	if err != nil {
+		return nil, err
+	}
+
+	bundle := &protobundle.Bundle{
+		MediaType: "application/vnd.dev.sigstore.bundle.v0.3+json",
```

Comment:
Do we have a constant we can use for this string?

Reply:
I didn't see one elsewhere in `sigstore`, `sigstore-go`, or `protobuf-specs`, but I moved it into a constant.